e2e/test/iothub/AuthenticationWithTokenRefreshDisposalTests.cs

```diff
@@ -0,0 +1,252 @@
+                amqpConnectionStatuses.Add(amqpConnectionStatusChange);
+
+                await deviceClient.OpenAsync().ConfigureAwait(false);
+                deviceClients.Add(deviceClient);
+            }
+
+            // Ensure that all clients are connected successfully, and the close and dispose the instances.
+            for (int i = 0; i < devicesCount; i++)
+            {
+                amqpConnectionStatuses[i].LastConnectionStatus.Should().Be(ConnectionStatus.Connected);
+
+                await deviceClients[i].CloseAsync();
+                deviceClients[i].Dispose();
+
+                amqpConnectionStatuses[i].LastConnectionStatus.Should().Be(ConnectionStatus.Disabled);
+            }
+        }
+
+        private class DeviceAuthenticationSasToken : DeviceAuthenticationWithTokenRefresh
+        {
+            private const string SasTokenTargetFormat = "{0}/devices/{1}";
+            private readonly IotHubConnectionStringBuilder _connectionStringBuilder;
+
+            public DeviceAuthenticationSasToken(
+                string connectionString)
+                : base(GetDeviceIdFromConnectionString(connectionString))
+            {
+                if (connectionString == null)
+                {
+                    throw new ArgumentNullException(nameof(connectionString));
+                }
+
+                _connectionStringBuilder = IotHubConnectionStringBuilder.Create(connectionString);
+            }
+
+            protected override Task<string> SafeCreateNewToken(string iotHub, int suggestedTimeToLive)
+            {
+                var builder = new SharedAccessSignatureBuilder
+                {
+                    Key = _connectionStringBuilder.SharedAccessKey,
+                    TimeToLive = TimeSpan.FromSeconds(suggestedTimeToLive),
+                };
+
+                if (_connectionStringBuilder.SharedAccessKeyName == null)
+                {
+                    builder.Target = string.Format(
+                        CultureInfo.InvariantCulture,
+                        SasTokenTargetFormat,
+                        iotHub,
+                        WebUtility.UrlEncode(DeviceId));
+                }
+                else
+                {
+                    builder.KeyName = _connectionStringBuilder.SharedAccessKeyName;
+                    builder.Target = _connectionStringBuilder.HostName;
+                }
+
+                return Task.FromResult(builder.ToSignature());
+            }
+
+            private static string GetDeviceIdFromConnectionString(string connectionString)
+            {
+                if (connectionString == null)
+                {
+                    throw new ArgumentNullException(nameof(connectionString));
+                }
+
+                var builder = IotHubConnectionStringBuilder.Create(connectionString);
+                return builder.DeviceId;
+            }
```

Comment:
These tests are doing a lot more than just checking for reusing the disposable created external to the SDK. Is that because we don't have tests that cover these mux scenarios? If they already exist, it would be nice simplify to just testing the reusability.

Reply:
We have tests that cover basic mux and pooling scenarios (all operations), and that cover fault injection when devices have opted for multiplexing (and pooling).
These new tests are designed to reuse the authentication method, while still testing that disposal on one multiplexed device doesn't cause cascading failures in the rest of the devices on the same tcp connection.

Reply:
This seems more explanatory. Can you just add this in comments?

Reply:
sure, I'll add this comment in